### PR TITLE
Async search: don't track fetch failures

### PR DESCRIPTION
**Pull request by javanna**
Fetch failures are currently tracked by AsyncSearchTask like ordinary shard failures. Though they should be treated differently or they end up causing weird scenarios like total=num_shards and successful=num_shards as the query phase ran fine yet the failed count would reflect the number of shards where fetch failed.

Given that partial results only include aggs for now and are complete even if fetch fails, we can ignore fetch failures in async search, as they will be anyways included in the response. They are in fact either received as a failure when all shards fail during fetch, or as part of the final response when only some shards fail during fetch.

We decided to make this change about 5 months ago [here](https://github.com/elastic/elasticsearch/pull/55758/#discussion_r421407483) but I only got to it now.

### CODE REVIEW

**Comment by javanna** [comment]
run elasticsearch-ci/bwc

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
     private final int totalShards;
     private final int skippedShards;
     private final Clusters clusters;
-    private final AtomicArray<ShardSearchFailure> shardFailures;
+    private final AtomicArray<ShardSearchFailure> queryFailures;
```
Comment (jimczi) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:38]: Can we keep the name as is for now. I understand that we want to call it `query` failures after this change but we may also want to provide a way to return partial top hits in the future. My point being that we might reintroduce fetch failures when they make sense.

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    public static final class SubFetchPhasePlugin extends Plugin implements SearchPlugin {
+        @Override
+        public List<FetchSubPhase> getFetchSubPhases(FetchPhaseConstructionContext context) {
+            return Collections.singletonList(searchContext -> new FetchSubPhaseProcessor() {
+                @Override
+                public void setNextReader(LeafReaderContext readerContext) {}
+
+                @Override
+                public void process(FetchSubPhase.HitContext hitContext) {
+                    if (hitContext.hit().getId().startsWith("boom")) {
```
Comment (jimczi) [inline x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java:126]: nit: you could derive it from the index name instead ?

**Review by jimczi — APPROVED** [review]
I left one comment around naming and a future direction. The change looks good to me, I don't think we document the behavior anywhere so that would be nice to have a follow up to explain partial responses and what failure or success means in that context.

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java (hunk 2 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java:126]: that is what I thought too but it turns out the index is null, no idea why.

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java (hunk 1 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:38]: this is not a huge deal, but can we not rename it back to shardFailures in the future once we introduce a way to return partial top hits?

Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:38]: As you want, that's not a big deal I agree but I wanted to raise the awareness that returning partial top hits is still in the air.

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java:126]: do you mean `hitContext.hit().getIndex()` is null ?

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java (hunk 1 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:38]: Thanks, I am aware, and I probably would have made the same comment if you had made this same change :)

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java (hunk 2 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java:126]: yes pretty much everything is null in the hit besides the id

**Comment by javanna** [comment]
thanks @jimczi !

**Pull request merged**